src/java/picard/sam/revertsam/RevertSam.java

```diff
@@ -240,37 +212,97 @@ protected int doWork() {
 
-                    public boolean filterOut(final SAMRecord first, final SAMRecord second) {
-                        throw new UnsupportedOperationException();
-                    }
-                };
-                readGroupToFormat.put(rg, QualityEncodingDetector.detect(QualityEncodingDetector.DEFAULT_MAX_RECORDS_TO_ITERATE, new FilteringIterator(reader.iterator(), filter), RESTORE_ORIGINAL_QUALITIES));
-                CloserUtil.close(reader);
+            if (discarded / (double) total > MAX_DISCARD_FRACTION) {
+                throw new PicardException("Discarded " + fmt.format(discardRate) + " which is above MAX_DISCARD_FRACTION of " + fmt.format(MAX_DISCARD_FRACTION));
             }
-            for (final SAMReadGroupRecord r : readGroupToFormat.keySet()) {
-                log.info("Detected quality format for " + r.getReadGroupId() + ": " + readGroupToFormat.get(r));
+        }
+
+        CloserUtil.close(in);
+        return 0;
+    }
+
+    private boolean isPresorted(final SAMFileHeader inHeader, final SortOrder sortOrder, final boolean sanitizing) {
+        return (inHeader.getSortOrder() == sortOrder) || (sortOrder == SortOrder.queryname && sanitizing);
+    }
+
+    /**
+     * Takes an individual SAMRecord and applies the set of changes/reversions to it that
+     * have been requested by program level options.
+     */
+    private void revertSamRecord(final SAMRecord rec) {
+        if (RESTORE_ORIGINAL_QUALITIES) {
+            final byte[] oq = rec.getOriginalBaseQualities();
+            if (oq != null) {
+                rec.setBaseQualities(oq);
+                rec.setOriginalBaseQualities(null);
             }
-            if (readGroupToFormat.values().contains(FastqQualityFormat.Solexa)) {
-                log.error("No quality score encoding conversion implemented for " + FastqQualityFormat.Solexa);
-                return -1;
+        }
+
+        if (REMOVE_DUPLICATE_INFORMATION) {
+            rec.setDuplicateReadFlag(false);
+        }
+
+        if (REMOVE_ALIGNMENT_INFORMATION) {
+            if (rec.getReadNegativeStrandFlag()) {
+                SAMRecordUtil.reverseComplement(rec);
+                rec.setReadNegativeStrandFlag(false);
+            }
+
+            // Remove all alignment based information about the read itself
+            rec.setReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+            rec.setAlignmentStart(SAMRecord.NO_ALIGNMENT_START);
+            rec.setCigarString(SAMRecord.NO_ALIGNMENT_CIGAR);
+            rec.setMappingQuality(SAMRecord.NO_MAPPING_QUALITY);
+
+            if (!rec.getReadUnmappedFlag()) {
+                rec.setInferredInsertSize(0);
+                rec.setNotPrimaryAlignmentFlag(false);
+                rec.setProperPairFlag(false);
+                rec.setReadUnmappedFlag(true);
+
+            }
+
+            // Then remove any mate flags and info related to alignment
+            if (rec.getReadPairedFlag()) {
```

Comment:
Yes. I agree. Mate information is mapping information and should be removed
during revert regardless of how the bam got to its sad state.

On Thu, May 5, 2016 at 3:15 PM, Nils Homer <notifications@github.com> wrote:

> In src/java/picard/sam/revertsam/RevertSam.java
> <https://github.com/broadinstitute/picard/pull/531#discussion_r62238942>:
>
> > +            // Remove all alignment based information about the read itself
> > +            rec.setReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
> > +            rec.setAlignmentStart(SAMRecord.NO_ALIGNMENT_START);
> > +            rec.setCigarString(SAMRecord.NO_ALIGNMENT_CIGAR);
> > +            rec.setMappingQuality(SAMRecord.NO_MAPPING_QUALITY);
> > +
> > +            if (!rec.getReadUnmappedFlag()) {
> > +                rec.setInferredInsertSize(0);
> > +                rec.setNotPrimaryAlignmentFlag(false);
> > +                rec.setProperPairFlag(false);
> > +                rec.setReadUnmappedFlag(true);
> > +
> > +            }
> > +
> > +            // Then remove any mate flags and info related to alignment
> > +            if (rec.getReadPairedFlag()) {
>
> Some software *incorrectly* has mate information even when the read
> paired flag is not set, and it would be great if RevertSam removed that
> mate information. If the goal of this tool is to have a clean and valid BAM
> with no mapping information in the output, I don't see what harm removing
> all this information regardless of whether or not the getReadPairedFlag
> is set. @yfarjoun <https://github.com/yfarjoun> would you confirm my
> reasoning?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/531/files/99dbe60dd4f272e090c94f29ac77624d70dc50bc#r62238942>
>